### PR TITLE
Pod topology hints

### PR DESCRIPTION
**Pull request by pablitoergosum**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by pablitoergosum** [comment]
This should not be merged before #28

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 1]
```diff
@@ -238,6 +238,63 @@ func (p *staticPolicy) RemoveContainer(s state.State, podUID string, containerNa
...
+	if v1qos.GetPodQOS(pod) != v1.PodQOSGuaranteed {
+		return nil
+	}
+
+	reqRsrcs, err := p.getPodRequestedResources(pod)
+	if err != nil {
+		klog.Error(err.Error())
+		return nil
+	}
+
```
Comment (pablitoergosum) [inline pkg/kubelet/cm/memorymanager/policy_static.go:294]: @cynepco3hahue This function does not check extraordinary cases, thoroughly examined in ``GetTopologyHints()``. The question is: should it be copied from the function mentioned and looped over all containers or there's a better solution to the problem?

**Comment by k-wiatrzyk** [comment]
Just noticed: need to write unit tests for two new funcs. (GetPodTopologyHints(), GetPodRequestedResources)

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 1 above)
Comment (cynepco3hahue) [reply pkg/kubelet/cm/memorymanager/policy_static.go:294]: IMHO we should do the same short circuit that done under the `GetTopologyHints`, I think we can move the code:
```golang
		if len(containerBlocks) != len(requestedResources) {
			klog.Errorf("[memorymanager] The number of requested resources by the container %s differs from the number of memory blocks", container.Name)
			return nil
		}

		for _, b := range containerBlocks {
			if _, ok := requestedResources[b.Type]; !ok {
				klog.Errorf("[memorymanager] Container %s requested resources do not have resource of type %s", container.Name, b.Type)
				return nil
			}

			if b.Size != requestedResources[b.Type] {
				klog.Errorf("[memorymanager] Memory %s already allocated to (pod %v, container %v) with different number than request: requested: %d, allocated: %d", b.Type, pod.UID, container.Name, requestedResources[b.Type], b.Size)
				return nil
			}

			containerNUMAAffinity, err := bitmask.NewBitMask(b.NUMAAffinity...)
			if err != nil {
				klog.Errorf("[memorymanager] failed to generate NUMA bitmask: %v", err)
				return nil
			}

			klog.Infof("[memorymanager] Regenerating TopologyHints, %s was already allocated to (pod %v, container %v)", b.Type, pod.UID, container.Name)
			hints[string(b.Type)] = append(hints[string(b.Type)], topologymanager.TopologyHint{
				NUMANodeAffinity: containerNUMAAffinity,
				Preferred:        true,
			})
```

under the separate function and re-use it.

**Review by cynepco3hahue — COMMENTED** [review]
I left some comments but except this one the code under the static policy lgtm
@k-wiatrzyk Give me know when memory manager unittests will be ready, I will review it.

File: pkg/kubelet/cm/memorymanager/policy_static.go [hunk 2]
```diff
@@ -220,9 +220,9 @@ func (p *staticPolicy) RemoveContainer(s state.State, podUID string, containerNa
 			// the reserved memory smaller than the amount of the memory that should be released
 			// release as much as possible and move to the next node
 			if nodeResourceMemoryState.Reserved < releasedSize {
+				releasedSize -= nodeResourceMemoryState.Reserved
```
[the remark this answers is not in the document]
Comment (k-wiatrzyk) [reply pkg/kubelet/cm/memorymanager/policy_static.go:223]: Well, subtracting at the end after setting `nodeResourceMemoryState.Reserved = 0` caused no decrementing of `releaseSize` and IIRC it caused problems with tests when more than one pod is used.

[the remark this answers is not in the document]
Comment (cynepco3hahue) [reply pkg/kubelet/cm/memorymanager/policy_static.go:223]: oh my bad

[the remark this answers is not in the document]
Comment (pablitoergosum) [reply pkg/kubelet/cm/memorymanager/policy_static.go:223]: Oh, this is not relevant here - this is part of another PR, #28.
I should've explained it before. The change I made was simply adding ``GetPodTopologyHints()`` and ``getPodRequestedResources``.

File: pkg/kubelet/cm/memorymanager/policy_static.go (hunk 1 above)
Comment (pablitoergosum) [reply pkg/kubelet/cm/memorymanager/policy_static.go:332]: @cynepco3hahue Thanks a lot, it seems the most reasonable in that case. Just pushed the code.

**Comment by cynepco3hahue** [comment]
@pablitoergosum Can you please rebase once we will merge MemoryManager unittests and put a little more description under commits messages and probably squash it to the single commit(adding the support for the GetPodTopologyHints). Because we will need to have more explanatory commit history

**Comment by pablitoergosum** [comment]
@cynepco3hahue Of course, I will do it with pleasure.

**Pull request merged**